src/config/preprocess-config.ts

```diff
@@ -0,0 +1,55 @@
+import fastGlob from 'fast-glob';
+import path from 'path';
+import { Config, RuleUsage, ConfigFile, RuleFn, ConfigRules, ConfigAPI } from './../types';
+import { getDirResolver } from '../utils';
+
+const getRuleUsages = (
+    rulesFn: RuleFn,
+    root: string,
+    /* Cache is mutated */
+    cache: { packages?: string[] } = {},
+    packageGlobs: string[] = ['*/**'],
+): RuleUsage | RuleUsage[] => {
+    const pkgJsonGlobs = packageGlobs.map(p => `${p}/package.json`);
+    if (!cache.packages) {
+        const resolvedPackages: string[] = fastGlob
+            .sync<string>([...pkgJsonGlobs, '!**/node_modules/**'], {
+                cwd: root,
+            })
+            .map(pkgJsonPath => path.dirname(pkgJsonPath));
+        cache.packages = resolvedPackages;
+    }
+    return rulesFn({ packages: cache.packages });
+};
+
+const transformRules = (config: ConfigAPI): ConfigRules => {
+    const cache = {};
```

Comment:
Sorry, this code is a bit complicated to grasp. Could you, please, simplify it?
I suggest populating `cache` outside of `getRuleUsages`.
You could check, if there are function rules first. If that is the case, populate the cache and use it in `reduce` below.